# Students Performance Indicator



## Life cycle of Machine Learning Project

+ Uderstanding the Problem Statement
+ Data Collection
+ Data Checks to Perform
+ Exploratory Data Analysis
+ Data Preprocessing
+ Model Training
+ Choose best Model


1.) Problem Statement
* This project understands how the student Perfomance (test scores) is affected by other variables Gender, Ethnicity, Parental level of education,Lunch and Test preparation course.

2.) Data Collection
+ Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams
+ The data consists of 8 columns and 1000 rows.

3.) Import Data and required packages


## Importing Pandas, Numpy,Matplotlib, Seaborn, Warring library.

In [9]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

Matplotlib is building the font cache; this may take a moment.


## Import the CSV Data as Pandas DataFrame

In [12]:
df = pd.read_csv("data\StudentsPerformance.csv")

Show Top 5 Records

In [14]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

Shape of the Dataset

In [15]:
df.shape

(1000, 8)

# 2.2 Dataset Information

+ gender: Sex of students-> (Male/Female)
+ race/ethnicity: ethnicity of students->(Group A,B,C,D,E)
+ parental level of education: parent's final year->(bachelor;s degree, some college, masters degree, associate's degree, high  school)
+ lunch: having lunch before test(standard or free/reduced)
+ test preparation course: complete or not complete before test
+ maths score
+ reading score
+ writing score




# 3. Datat Checks to Perform

+ Check Missing Values
+ Check Duplicates
+ Check Data types
+ Check the Number of uniques values in each columns
+ Check statistics of the Dataset
+ Check various categories present in the different categorical column


# 3.1 Checking Missing Values

In [16]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

There are no missing values in the Dataset

# 3.2 Check Duplicates

In [17]:
df.duplicated().sum()

np.int64(0)

There are no duplicates value in the Dataset.

# 3.3 Check Data Type

In [19]:
# Check Null and DTypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


# 3.4 Checking the Number of Unique Values of each columns

In [20]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

# 3.5 check Statistics of the Data Set

In [21]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

insight

+ from the above desription of numerical data, all means are very close to each other, between 66 and 68.08;
+ All standard deviation are also very close to each other - between 14.6 and 15.19
* While there is a minimum score 0 for maths, for writing minimum is much higher = 10, and for reading much more higher = 17


# 3.7 Exploring Data

In [23]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [26]:
print("categories in 'gender' variable:  ", end= " ")
print(df['gender'].unique())

print("categories in 'race/ethnicity' variable:  ", end= " ")
print(df['race/ethnicity'].unique())

print("categories in 'parental level of education	' variable:  ", end= " ")
print(df['parental level of education'].unique())


print("categories in 'lunch' variable:  ", end= " ")
print(df['lunch'].unique())


print("categories in 'test preparation course' variable:  ", end= " ")
print(df['test preparation course'].unique())


#print("categories in 'math score' variable:  ", end= " ")
#print(df['math score'].unique())


#print("categories in 'reading score' variable:  ", end= " ")
#print(df['reading score'].unique())

#print("categories in 'writing score' variable:  ", end= " ")
#print(df['writing score'].unique())

categories in 'gender' variable:   ['female' 'male']
categories in 'race/ethnicity' variable:   ['group B' 'group C' 'group A' 'group D' 'group E']
categories in 'parental level of education	' variable:   ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
categories in 'lunch' variable:   ['standard' 'free/reduced']
categories in 'test preparation course' variable:   ['none' 'completed']


In [27]:
#define numerical and categorical columns
numeric_features = [feature for feature in df.columns if df[feature].dtype != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']

#print columns
print("We have {} numerical features : {}".format(len(numeric_features), numeric_features))
print("nWe have {} categorical features : {}".format(len(categorical_features), categorical_features))


We have 3 numerical features : ['math score', 'reading score', 'writing score']
nWe have 5 categorical features : ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']


# 3.8 Adding Columns for "Total Score" and "Average"

In [33]:
df["total score"] = df["math score"] + df["reading score"] + df["writing score"]
df["average"] = df["total score"]
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total score  average score  average  
0          218            218      218  
1          247            247      247  
2          278            278      278  
3          148            148      148  
4          229            229      229

In [34]:
reading_full = df[df["reading score"] == 100]["average"].count()
writing_full = df[df["writing score"] == 100]["average"].count()
math_full = df[df["math score"] == 100]["average"].count()

print(f"Number of students with full marks in Maths: {math_full}")
print(f"Number of students with full marks in Writing: {writing_full}")
print(f"Number of students with full marks in Reading: {reading_full}")

Number of students with full marks in Maths: 7
Number of students with full marks in Writing: 14
Number of students with full marks in Reading: 17


In [35]:
reading_less_20 = df[df["reading score"] <= 20]["average"].count()
writing_less_20 = df[df["writing score"] <= 20]["average"].count()
math_less_20 = df[df["math score"] <= 20]["average"].count()

print(f"Number of students with less than 20 marks in Maths: {math_less_20}")
print(f"Number of students with less than 20 marks in Writing: {writing_less_20}")
print(f"Number of students with less than 20 marks in Reading: {reading_less_20}")

Number of students with less than 20 marks in Maths: 4
Number of students with less than 20 marks in Writing: 3
Number of students with less than 20 marks in Reading: 1


Insight

+ From the above values, we get students have performed worst in Maths
+ Best performance is in reading section



# 4. Exploring Data(Visualization)

## 4.1 Visualize average score distribution to make some conclusion.


+ Histogram
+ Kernel Distribution Function(KDE)

## 4.1.1 Histogram & KDE

In [37]:
fig, axs = plt.subplots(1,2, figsize=(15,7))
plt.subplot(121)
sns.histplot(data=df, x="average", kde=True, color="g")
plt.subplot(122)
sns.histplot(data=df, x="average",kde=True,hue="gender")

<Axes: xlabel='average', ylabel='Count'>